cli/src/commands/rebase.rs

```diff
@@ -443,6 +453,127 @@ fn rebase_revision(
     }
 }
 
+/// Extracts `target_commits` from the graph by rebasing its descendants onto
+/// its parents. This assumes that `target_commits` can be rewritten.
+/// Returns a map from the original commit IDs to the rewritten commit IDs, and
+/// the roots of the `target_commits` set, using the rewritten commit ID if
+/// applicable.
+fn extract_commits(
+    tx: &mut WorkspaceCommandTransaction,
+    settings: &UserSettings,
+    target_commits: &IndexSet<Commit>,
+) -> Result<(HashMap<CommitId, CommitId>, IndexSet<CommitId>), CommandError> {
+    // Commits in the target set should only have other commits in the set as
+    // parents, except the roots of the set, which persist their original
+    // parents.
+    let mut new_target_parents: HashMap<CommitId, Vec<CommitId>> = HashMap::new();
+    for commit in target_commits.iter().rev() {
+        // The roots of the set will not have any parents found in `new_target_parents`,
+        // and will be stored in `new_target_parents` as an empty vector.
+        let mut new_parents = vec![];
+        for old_parent in commit.parent_ids() {
+            if new_target_parents.contains_key(old_parent) {
+                new_parents.push(old_parent.clone());
+            }
+        }
+        new_target_parents.insert(commit.id().clone(), new_parents);
+    }
+
```

Comment:
Instead of rebasing multiple disconnected subgraphs separately (as we do in `parallelize`), shouldn't we rebase the target commits as one subgraph (with non-target commits omitted)? I think `hg rebase -r` works that way.

In the following example copied from the test, `rebase -r 'd|g'` will create `(dest) <- d <- g` chain instead of `(dest) <- {d, g}`.
```
    │ ◉  h
    │ ◉  g
    ├─╯
    ◉    f
    ├─╮
    │ ◉  e
    │ ◉  d
```

I don't have a strong opinion, but just asking because the underlying logic would be quite different.

Reply:
Yes, I was going to ask the same. That's what Mercurial does. I remember being a bit surprised about that when I learned about it, but I think it's also usually been convenient. It makes sense to me that `rebase -r` says that "the commits in the set are actually independent of the descendants but still preserve internal dependencies"  (i.e. the opposite of `parallelize`). What do you think?

Reply:
I think persisting ancestral relationships makes sense; I've updated the code to do this by getting the connected graph of commits being rebased, and finding ancestors which are in the graph and in the target set.